hibernate-core/src/main/java/org/hibernate/generator/values/AbstractGeneratedValuesMutationDelegate.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.generator.values;
+
+import org.hibernate.dialect.Dialect;
+import org.hibernate.generator.EventType;
+import org.hibernate.generator.values.internal.GeneratedValuesHelper;
+import org.hibernate.generator.values.internal.GeneratedValuesMappingProducer;
+import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.sql.results.jdbc.spi.JdbcValuesMappingProducer;
+
+/**
+ * @author Marco Belladelli
+ */
+public abstract class AbstractGeneratedValuesMutationDelegate implements GeneratedValuesMutationDelegate {
+	protected final EntityPersister persister;
+	private final EventType timing;
+	protected final GeneratedValuesMappingProducer jdbcValuesMappingProducer;
+
+	public AbstractGeneratedValuesMutationDelegate(EntityPersister persister, EventType timing) {
+		this( persister, timing, false );
+	}
+
+	public AbstractGeneratedValuesMutationDelegate(EntityPersister persister, EventType timing, boolean inferredKeys) {
+		this.persister = persister;
+		this.timing = timing;
+		this.jdbcValuesMappingProducer = GeneratedValuesHelper.createMappingProducer(
+				persister,
+				timing,
+				supportsArbitraryValues() && !inferredKeys,
```

Comment:
It's kind of dangerous to use this method as subclasses implement it based on state that is initialized after this constructor has ran. I would suggest you add an additional constructor argument instead of calling this method. Maybe you can even drop the method then.

Reply:
I've created a new constructor as you suggested.

While investigating if I could drop this method, I've modified the reading of generated values (i.e. `GeneratedValuesHelper#getGeneratedValues`) to use the mapping producer's list of result builders, like you suggested for delegate constructors. I had to cast to the explicit class, but after that I could use the list of generated attributes from there instead of re-building it again. The method is still used in `GeneratedValuesProcessor` and tests, and I believe it might be useful so I think we can keep it around.